src/main/java/org/terasology/web/io/JsonSession.java

```diff
@@ -0,0 +1,92 @@
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonSyntaxException;
+import org.terasology.config.Config;
+import org.terasology.entitySystem.entity.EntityManager;
+import org.terasology.identity.storageServiceClient.BigIntegerBase64Serializer;
+import org.terasology.web.EngineRunner;
+import org.terasology.web.authentication.AuthenticationFailedException;
+import org.terasology.web.authentication.AuthenticationHandshakeHandler;
+import org.terasology.web.authentication.AuthenticationHandshakeHandlerImpl;
+import org.terasology.web.authentication.ClientAuthenticationMessage;
+import org.terasology.web.authentication.HandshakeHello;
+import org.terasology.web.client.HeadlessClient;
+import org.terasology.web.client.HeadlessClientFactory;
+
+import java.math.BigInteger;
+
+public class JsonSession {
+
+    private static final Gson GSON = new GsonBuilder()
+            .registerTypeHierarchyAdapter(BigInteger.class, BigIntegerBase64Serializer.getInstance())
+            .registerTypeAdapter(byte[].class, ByteArrayBase64Serializer.getInstance())
+            .create();
+
+    private final AuthenticationHandshakeHandler authHandler;
+    private final HeadlessClientFactory headlessClientFactory;
+
+    private HeadlessClient client;
+
+    JsonSession(AuthenticationHandshakeHandler authHandler, HeadlessClientFactory headlessClientFactory) {
+        this.authHandler = authHandler;
+        this.headlessClientFactory = headlessClientFactory;
+    }
+
+    public JsonSession() {
+        this(new AuthenticationHandshakeHandlerImpl(EngineRunner.getContext().get(Config.class).getSecurity()),
+                new HeadlessClientFactory(EngineRunner.getContext().get(EntityManager.class)));
+    }
+
+    public boolean isAuthenticated() {
+        return client != null;
+    }
+
+    public ActionResult initAuthentication() {
+        if (isAuthenticated()) {
+            return new ActionResult(ActionResult.Status.UNAUTHORIZED, "Already authenticated");
+        }
+        HandshakeHello serverHello = authHandler.initServerHello();
+        return new ActionResult(GSON.toJsonTree(serverHello));
+    }
+
+    public ActionResult finishAuthentication(JsonElement clientMessage) {
+        if (isAuthenticated()) {
+            return new ActionResult(ActionResult.Status.UNAUTHORIZED, "Already authenticated");
+        }
+        try {
+            ClientAuthenticationMessage clientAuthentication = GSON.fromJson(clientMessage, ClientAuthenticationMessage.class);
+            byte[] serverVerification = authHandler.authenticate(clientAuthentication);
+            String clientId = clientAuthentication.getClientHello().getCertificate().getId();
+            client = headlessClientFactory.connectNewHeadlessClient(clientId);
+            return new ActionResult(GSON.toJsonTree(serverVerification));
+        } catch (NullPointerException | JsonSyntaxException ex) {
```

Comment:
Not a debug leftover, the reason it's there is that an invalid JSON message sent by the client could miss some field (for example the signature) and thus the deserialized object will have those fields set to `null`, resulting in NPEs during the verification process.
Anyway, I agree that catching unchecked exception is not that good. I think I'll fix this by adding a validation method in ClientAuthenticationMessage which throws a custom exception if some mandatory fields are null, like I did for the deserialization of WebSocket resource request messages.